stellargraph/mapper/mini_batch_node_generators.py

```diff
@@ -0,0 +1,361 @@
+    """
+    A data generator for use with ClusterGCN models on homogeneous graphs, [1].
+
+    The supplied graph G should be a StellarGraph object that is ready for
+    machine learning. Currently the model requires node features to be available for all
+    nodes in the graph.
+    Use the :meth:`flow` method supplying the nodes and (optionally) targets
+    to get an object that can be used as a Keras data generator.
+
+    This generator will supply the features array and the adjacency matrix to a
+    mini-batch Keras graph ML model.
+
+    [1] `W. Chiang, X. Liu, S. Si, Y. Li, S. Bengio, C. Hsieh, 2019 <https://arxiv.org/abs/1905.07953>`_.
+
+    For more information, please see the ClusterGCN demo:
+        `<https://github.com/stellargraph/stellargraph/blob/master/demos/>`_
+
+    Args:
+        G (StellarGraphBase): a machine-learning StellarGraph-type graph
+        k (int): The number of clusters if parameter `clusters` is None. Otherwise it is ignored.
+        lam (float): The mixture coefficient for adjacency matrix normalisation.
+        clusters (list): a list of lists of node IDs such that each list corresponds to a cluster of nodes
+            in G. The clusters should be non-overlapping. If None, the G is clustered into k clusters.
+        name (str): an optional name of the generator
+    """
+
+    def __init__(self, G, k=1, q=1, lam=0.1, clusters=None, name=None):
+
+        if not isinstance(G, StellarGraphBase):
+            raise TypeError("Graph must be a StellarGraph object.")
+
+        self.graph = G
+        self.name = name
+        self.k = k
+        self.q = q  # The number of clusters to sample per mini-batch
+        self.lam = lam
+        self.clusters = clusters
+
+        if clusters:
+            self.k = len(clusters)
+
+        # Some error checking on the given parameter values
+        if lam < 0 or lam > 1 or not isinstance(lam, float):
+            raise ValueError(
+                "{}: lam must be a float in the range [0, 1].".format(
+                    type(self).__name__
+                )
+            )
+
+        if q <= 0 or not isinstance(q, int):
+            raise ValueError(
+                "{}: q must be a positive integer.".format(type(self).__name__)
+            )
+
+        if k <= 0 or not isinstance(k, int):
+            raise ValueError(
+                "{}: k must be a positive integer.".format(type(self).__name__)
+            )
+
+        if k % q != 0:
+            raise ValueError(
+                "{}: k must be exactly divisible by q.".format(type(self).__name__)
+            )
+
+        # Check if the graph has features
+        G.check_graph_for_ml()
+
+        self.node_list = list(G.nodes())
+
+        # We need a schema to check compatibility with ClusterGCN
+        self.schema = G.create_graph_schema(create_type_maps=True)
+
+        # Check that there is only a single node type
+        if len(self.schema.node_types) > 1:
+            raise TypeError(
```

Comment:
`ValueError` here.

Reply:
This is probably up to debate as the problem is the the graph has more than one node types. Our FullBatchNodeGenerator that we use for our GCN implementation return TypeError for the same check, so for consistency, I will not be changing it.

Reply:
Yes, but the type of `self.schema.node_types` is not the problem here (it is presumably a list), it is the value of that variable that is wrong, hence `ValueError`. The use of `TypeError` has nothing to do with node types. 

And I have to say that it is a facile argument to not make a suggested correction just because the same problem occurs in other places! We should be making our (collective `our`) code cleaner whenever we touch it, not leave the code smell intact.

Reply:
I've changed it to ValueError. I don't think this is issue is important enough to spend too much of our time arguing the correct exception to return.